### PR TITLE
Data Efficient Image Transformer (DEIT)

### PR DESCRIPTION
**Pull request by sigeisler**
# Description

Implementation of the Data-Efficient Image Transformer (DEIT) without distillation but including new augmentations:
- ColorJitter
- Stochastic Depth for the Transformer Encoder
- RandomErasing
- Mixup and Cutmix
- Repeated Augmentation

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Documentation update
- [ ] New feature (non-breaking change which adds functionality)
- [ ] A new research paper code implementation

## Tests

I ran several training procedures (varying the augmentation techniques) and added basic unit tests to the new augmentations. More details on the experiments and models will follow.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [ ] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by sigeisler** [comment]
@xianzhidu some open points:
1. Currently, I duplicated `official/nlp/keras_nlp/layers/transformer_encoder_block.py` to implement the stochastic depth. Should this class stay in the vision module in the long run or should we implement stochastic depth also for the nlp applications?
1. Is it fine that I modified [`official/core/input_reader.py`](https://github.com/tensorflow/models/pull/10187/files#diff-a1cd9c86fa47fb0abaa7a2df9902d301a001ca2fbe3b5fd25cff49d08d67b3bd) (as well as `official/core/config_definitions.py`) because of repeated augment? Or should I copy/move it to `official/vision/beta`?
1. I think it makes sense to remove at least some of the configs in [`official/vision/beta/projects/vit`](https://github.com/tensorflow/models/pull/10187/files#diff-f6da9b50936d68237f71bb56d25b2c8926201511797e3667db187d4eb6170296). What do you say?

**Comment by sigeisler** [comment]
@googlebot I signed it!

File: official/vision/beta/configs/common.py [hunk 1]
```diff
@@ -42,6 +43,30 @@ class AutoAugment(hyperparams.Config):
   translate_const: float = 250
 
 
+@dataclasses.dataclass
+class RandomErasing(hyperparams.Config):
```
Comment (xianzhidu) [inline official/vision/beta/configs/common.py:47]: Are all the arguments need to be exposed to users? Do you know which ones are useful and which ones are not? We can hide arguments as default values in the implementation.

File: official/vision/beta/ops/augment.py [hunk 2]
```diff
@@ -1267,3 +1306,240 @@ def distort(self, image: tf.Tensor) -> tf.Tensor:
...
+
+          image = _fill_rectangle(image, center_width, center_height,
+                                  half_width, half_height, replace=None)
+
+          break
+
+    return image
+
+
+class MixupAndCutmix:
```
Comment (xianzhidu) [inline official/vision/beta/ops/augment.py:1406]: Can you separate them to two classes?

File: official/vision/beta/projects/vit/modeling/layers/vit_transformer_encoder_block.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+"""Keras-based TransformerEncoder block layer."""
+
+import tensorflow as tf
+
+from official.vision.beta.modeling.layers.nn_layers import StochasticDepth
+
+
+@tf.keras.utils.register_keras_serializable(package="Vision")
+class TransformerEncoderBlock(tf.keras.layers.Layer):
```
Comment (xianzhidu) [inline official/vision/beta/projects/vit/modeling/layers/vit_transformer_encoder_block.py:23]: Can you simply subclass the original implementation then add necessary changes?

File: official/vision/beta/configs/common.py [hunk 4]
```diff
@@ -42,6 +43,30 @@ class AutoAugment(hyperparams.Config):
...
+  max_area: float = 1 / 3
+  min_aspect: float = 0.3
+  max_aspect = None
+  min_count = 1
+  max_count = 1
+  trials = 10
+
+
+@dataclasses.dataclass
+class MixupAndCutmix(hyperparams.Config):
```
Comment (xianzhidu) [inline official/vision/beta/configs/common.py:60]: Can you separate Mixup and Cutmix?

**Review by xianzhidu — COMMENTED** [review]
Awesome work! Thanks Simon.

File: official/vision/beta/ops/augment.py (hunk 2 above)
Comment (sigeisler) [reply official/vision/beta/ops/augment.py:1406]: The only difficulty in separating Cutmix and Mixup is located in [`distort`](https://github.com/tensorflow/models/pull/10187/files#diff-43bd3b3bd437704043e9d5bfa065c61b03e1488f5e9366e8820fe7e8b6b23caeR1473). Here, I randomly choose either Mixup or Cutmix. 

Hence, if I separate them into two classes I would need to either (1) move this "choose on of" feature to [the `image_classification` task](https://github.com/tensorflow/models/pull/10187/files#diff-5abe8251e133688b5f899bf7b14f718f98bb82fc1bc905ce28b51e393d3573b0R112) or (2) to add a "wrapper" here in the `augment.py`. What do you think is the best solution?

File: official/vision/beta/configs/common.py (hunk 1 above)
Comment (sigeisler) [reply official/vision/beta/configs/common.py:47]: I suggest having at least `probability` (the probability to erase some part of the image) and `max_count` (number of rectangles to "erase") configurable. On the other hand, I did not experiment a lot with those parameters. Hence, this is simply my best guess.

Comment (xianzhidu) [reply official/vision/beta/configs/common.py:47]: Let's keep them as what they are now. We can revisit this part later if we have a better understanding of the importance of each param.

File: official/vision/beta/ops/augment.py (hunk 2 above)
Comment (xianzhidu) [reply official/vision/beta/ops/augment.py:1406]: Let's keep it as it is in the first version.

**Review by xianzhidu — APPROVED** [review]

File: official/vision/beta/projects/vit/configs/image_classification.py [hunk 5]
```diff
@@ -81,630 +81,14 @@ class ImageClassificationTask(cfg.TaskConfig):
     image_classification.ImageClassificationTask)
 
 
-@exp_factory.register_config_factory('deit_imagenet_pretrain_noaug')
-def image_classification_imagenet_deit_imagenet_pretrain_noaug() -> cfg.ExperimentConfig:
+@exp_factory.register_config_factory('deit_imagenet_pretrain_nodistillation')
```
Comment (xianzhidu) [inline official/vision/beta/projects/vit/configs/image_classification.py:84]: just call it `deit_imagenet_pretrain`

**Review by xianzhidu — APPROVED** [review]
Thanks Simon!

**Review by xianzhidu — DISMISSED** [review]

File: official/vision/beta/projects/vit/configs/image_classification.py (hunk 5 above)
Comment (sigeisler) [reply official/vision/beta/projects/vit/configs/image_classification.py:84]: Done

**Comment by sigeisler** [comment]
I have just checked that I am able to reproduce the results even after the latest changes of the code

**Review by jaeyounkim — APPROVED** [review]

**Comment by jaeyounkim** [comment]
Please create a new PR after fixing issues.

**Pull request closed without merging**